### PR TITLE
feat(#69): distraction list

### LINKED ISSUE #69: Add a Screen to see a list of Distraction suggestion.

### PR DESCRIPTION
**Pull request by lipefree**
Closes #69.

Changed where made in the service for distractions to not use callbacks except for MapsActivity.

### CODE REVIEW

File: app/src/androidTest/java/com/github/studydistractor/sdp/procrastinationActivity/ProcrastinationActivityActivityViewModelTest.kt [hunk 1]
```diff
@@ -4,23 +4,16 @@ import org.junit.Assert.*
 import org.junit.Test
 
 class ProcrastinationActivityActivityViewModelTest {
-    val p = ProcrastinationActivityActivityViewModel()
-    val name = "salut"
-    val description = "bonjour"
+    val p = DistractionViewModel()
```
Comment (AthenaPK) [inline app/src/androidTest/java/com/github/studydistractor/sdp/procrastinationActivity/ProcrastinationActivityActivityViewModelTest.kt:7]: Maybe consider a more descriptive name than `p`.

File: app/src/androidTest/java/com/github/studydistractor/sdp/procrastinationActivity/ProcrastinationActivityActivityTest.kt
Comment (AthenaPK) [inline app/src/androidTest/java/com/github/studydistractor/sdp/procrastinationActivity/ProcrastinationActivityActivityTest.kt:1]: It's better if the file name is changed to use `distraction` instead of `procrastination.

File: app/src/androidTest/java/com/github/studydistractor/sdp/procrastinationActivity/ProcrastinationActivityActivityViewModelTest.kt
Comment (AthenaPK) [inline app/src/androidTest/java/com/github/studydistractor/sdp/procrastinationActivity/ProcrastinationActivityActivityViewModelTest.kt:1]: It's better if the file name is changed to use `distraction` instead of `procrastination.

File: app/src/androidTest/java/com/github/studydistractor/sdp/ui/DistractionScreenTest.kt [hunk 2]
```diff
@@ -25,13 +24,13 @@ class DistractionScreenTest {
...
-            DistractionScreen()
+//            DistractionListScreen()
         }
     }
 
     @Test
     fun testScreenIsShown() {
-        composeRule.onNodeWithTag("distraction-screen__main-container").assertIsDisplayed()
-        composeRule.onNodeWithText("Distraction screen").assertIsDisplayed()
+//        composeRule.onNodeWithTag("distraction-screen__main-container").assertIsDisplayed()
```
Comment (AthenaPK) [inline app/src/androidTest/java/com/github/studydistractor/sdp/ui/DistractionScreenTest.kt:33]: If the commented out code is not needed anymore, feel free to remove it. (same for line 27)

File: app/src/main/java/com/github/studydistractor/sdp/MainActivity.kt [hunk 3]
```diff
@@ -7,9 +7,19 @@ import androidx.compose.foundation.layout.fillMaxSize
...
+import com.github.studydistractor.sdp.procrastinationActivity.ProcrastinationActivityService
 import com.github.studydistractor.sdp.ui.theme.StudyDistractorTheme
+import dagger.hilt.android.AndroidEntryPoint
+import javax.inject.Inject
 
-class MainActivity : AppCompatActivity() {
+@AndroidEntryPoint(AppCompatActivity::class)
+class MainActivity : Hilt_MainActivity() {
+    @Inject
+    lateinit var hi : HistoryInterface
```
Comment (AthenaPK) [inline app/src/main/java/com/github/studydistractor/sdp/MainActivity.kt:19]: Consider a more descriptive name, e.g. `history` or `historyInterface`.

File: app/src/main/java/com/github/studydistractor/sdp/StudyDistractorScreen.kt [hunk 4]
```diff
@@ -42,10 +45,12 @@ enum class StudyDistractorScreen(@StringRes val title: Int) {
 
 @OptIn(ExperimentalMaterial3Api::class)
 @Composable
-@Preview("Study Distractor App")
+//@Preview("Study Distractor App")
```
Comment (AthenaPK) [inline app/src/main/java/com/github/studydistractor/sdp/StudyDistractorScreen.kt:48]: if the preview annotation is not needed anymore, feel free to remove it.

File: app/src/main/java/com/github/studydistractor/sdp/maps/MapsActivity.kt [hunk 5]
```diff
@@ -83,7 +88,7 @@ class MapsActivity : AppCompatActivity(), OnMapReadyCallback {
         }
         marker?.tag = activity
         map?.setOnInfoWindowClickListener { activityMarker ->
-            val procActivity = activityMarker.tag as ProcrastinationActivity
+            val procActivity = activityMarker.tag as Distraction
```
Comment (AthenaPK) [inline app/src/main/java/com/github/studydistractor/sdp/maps/MapsActivity.kt:91]: We can change the name of this variable to `distractionActivity` to be consistent with the renaming.

File: app/src/main/java/com/github/studydistractor/sdp/procrastinationActivity/DistractionList.kt [hunk 6]
```diff
@@ -20,12 +20,12 @@ class DistractionList : AppCompatActivity() {
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
-        FireBaseProcrastinationActivityService().fetchProcrastinationActivities { procrastinationActivities ->
-            displayActivities(procrastinationActivities)
-        }
+//        FireBaseProcrastinationActivityService().fetchProcrastinationActivities { procrastinationActivities ->
```
Comment (AthenaPK) [inline app/src/main/java/com/github/studydistractor/sdp/procrastinationActivity/DistractionList.kt:23]: Feel free to remove this comment, if the code is not needed anymore.

File: app/src/main/java/com/github/studydistractor/sdp/procrastinationActivity/AddProcrastinationActivityActivity.kt
Comment (AthenaPK) [inline app/src/main/java/com/github/studydistractor/sdp/procrastinationActivity/AddProcrastinationActivityActivity.kt:1]: We can change the file name to use distraction instead of `procrastination, to be consistent with the renaming.

File: app/src/main/java/com/github/studydistractor/sdp/procrastinationActivity/FireBaseProcrastinationActivityService.kt
Comment (AthenaPK) [inline app/src/main/java/com/github/studydistractor/sdp/procrastinationActivity/FireBaseProcrastinationActivityService.kt:1]: We can change the file name to use distraction instead of `procrastination, to be consistent with the renaming. The same applies to the package name.

File: app/src/main/java/com/github/studydistractor/sdp/procrastinationActivity/ProcrastinationActivityService.kt
Comment (AthenaPK) [inline app/src/main/java/com/github/studydistractor/sdp/procrastinationActivity/ProcrastinationActivityService.kt:1]: We can change the file name to use distraction instead of `procrastination, to be consistent with the renaming.

File: app/src/main/java/com/github/studydistractor/sdp/ui/DistractionListScreen.kt [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+            Divider(
+                modifier = Modifier.padding(horizontal = 14.dp),
+                color = MaterialTheme.colors.onSurface.copy(alpha = 0.08f)
+            )
+        }
+    }
+}
+
+@Composable
+fun ProcrastinationLayout(activity : Distraction,
```
Comment (AthenaPK) [inline app/src/main/java/com/github/studydistractor/sdp/ui/DistractionListScreen.kt:37]: Change name to use `Distraction`.

**Review by AthenaPK — CHANGES_REQUESTED** [review]
Overall, really good work! :ok_hand: 
One thing that I would recommend is splitting your commits, so that there are multiple small commits, each reflecting a single logical change. This will make our code much more maintable (and make it easier to review as well). Right now, if I understand correctly, your commit includes the refactor to rename `procrastination` to `distraction`, adding a FakeHistoryModule, moving the distraction list and activity to screens, and adding a maps workaround. There is a short tutorial here: https://stackoverflow.com/questions/6217156/break-a-previous-commit-into-multiple-commits

File: app/src/main/java/com/github/studydistractor/sdp/maps/MapsActivity.kt [hunk 8]
```diff
@@ -83,7 +88,7 @@ class MapsActivity : AppCompatActivity(), OnMapReadyCallback {
...
             val intent = Intent(this, ProcrastinationActivityActivity::class.java)
```
Comment (AthenaPK) [inline app/src/main/java/com/github/studydistractor/sdp/maps/MapsActivity.kt]: Since the Distraction Screen was added, I guess that this should take us to there instead of the (probably deprecated) ProcrastinationActivityActivity.

File: app/src/main/java/com/github/studydistractor/sdp/procrastinationActivity/DistractionList.kt [hunk 9]
```diff
@@ -20,12 +20,12 @@ class DistractionList : AppCompatActivity() {
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
```
Comment (AthenaPK) [inline app/src/main/java/com/github/studydistractor/sdp/procrastinationActivity/DistractionList.kt:22]: Since this commit adds DistractionListScreen, is this file still needed?

File: app/src/main/java/com/github/studydistractor/sdp/distraction/DistractionActivity.kt [hunk 10]
```diff
@@ -22,20 +21,19 @@ const val EXTRA_ACTIVITY = "activity"
 /**
  * This activity displays a procrastination activity
  */
-class ProcrastinationActivityActivity : AppCompatActivity() {
+class DistractionActivity : AppCompatActivity() {
```
Comment (aizensoosuke) [inline app/src/main/java/com/github/studydistractor/sdp/distraction/DistractionActivity.kt:24]: Activity :(

File: app/src/main/java/com/github/studydistractor/sdp/distraction/DistractionActivity.kt [hunk 11]
```diff
@@ -22,20 +21,19 @@ const val EXTRA_ACTIVITY = "activity"
...
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
 
         //https://stackoverflow.com/questions/73019160/android-getparcelableextra-deprecated
         val procrastinationData = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
-            intent.getParcelableExtra(EXTRA_ACTIVITY, ProcrastinationActivity::class.java)
+            intent.getParcelableExtra(EXTRA_ACTIVITY, Distraction::class.java)
```
Comment (aizensoosuke) [inline app/src/main/java/com/github/studydistractor/sdp/distraction/DistractionActivity.kt:30]: Parcel :((

File: app/src/main/java/com/github/studydistractor/sdp/distraction/DistractionList.kt [hunk 12]
```diff
@@ -20,12 +20,9 @@ class DistractionList : AppCompatActivity() {
...
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
-        FireBaseProcrastinationActivityService().fetchProcrastinationActivities { procrastinationActivities ->
-            displayActivities(procrastinationActivities)
-        }
     }
 
-    private fun displayActivities(procrastinationActivities: List<ProcrastinationActivity>) {
+    private fun displayActivities(procrastinationActivities: List<Distraction>) {
         setContent {
```
Comment (aizensoosuke) [inline app/src/main/java/com/github/studydistractor/sdp/distraction/DistractionList.kt:26]: Why is this file still an activity :(

File: app/src/main/java/com/github/studydistractor/sdp/distraction/DistractionService.kt [hunk 13]
```diff
@@ -0,0 +1,16 @@
+package com.github.studydistractor.sdp.distraction
+
+import androidx.compose.runtime.snapshots.SnapshotStateList
+interface DistractionService {
+
+    /**
+     * Fetch the distractions from the database
+     */
+    fun fetchDistractions() : SnapshotStateList<Distraction>
```
Comment (aizensoosuke) [inline app/src/main/java/com/github/studydistractor/sdp/distraction/DistractionService.kt:9]: Why a SnapshotStateList instead of a regular MutableStateList? Also maybe [LazyColumn](https://developer.android.com/jetpack/compose/lists?hl=fr) are the way to go

File: app/src/main/java/com/github/studydistractor/sdp/ui/DistractionListScreen.kt [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+import androidx.compose.ui.unit.dp
+import com.github.studydistractor.sdp.distraction.Distraction
+import com.github.studydistractor.sdp.distraction.DistractionViewModel
+import com.github.studydistractor.sdp.distraction.DistractionService
+
+/**
+ *  Screen showing the list of current available distraction on the DistractionService
+ */
+@Composable
+fun DistractionListScreen(distractionService : DistractionService,
```
Comment (aizensoosuke) [inline app/src/main/java/com/github/studydistractor/sdp/ui/DistractionListScreen.kt:23]: I feel like the order of the arguments does not make sense: keep all events for the end?

File: app/src/main/java/com/github/studydistractor/sdp/ui/DistractionListScreen.kt [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+            )
+        }
+    }
+}
+
+/**
+ * Distraction layout
+ */
+@Composable
+fun DistractionLayout(activity : Distraction,
```
Comment (aizensoosuke) [inline app/src/main/java/com/github/studydistractor/sdp/ui/DistractionListScreen.kt:43]: This should definitely be a separate file

File: app/src/main/java/com/github/studydistractor/sdp/maps/MapsActivity.kt (hunk 8 above)
Comment (lipefree) [reply app/src/main/java/com/github/studydistractor/sdp/maps/MapsActivity.kt]: I can't juste replace with screens, because MapsAcitivty is still not implemented screens.it is not straighforward for me to convert this. But once MapActivity is implemented with screens it will be straighforward to change. Or at least I don't know any way of doing it right now and I tried to do it

File: app/src/androidTest/java/com/github/studydistractor/sdp/DistractionActivityTest.kt [hunk 16]
```diff
@@ -44,7 +44,7 @@ inline fun <reified A: Activity> ComposeTestRule.launch(
  * See [testing documentation](http://d.android.com/tools/testing).
  */
 @RunWith(AndroidJUnit4::class)
-class ProcrastinationActivityActivityTest {
+class DistractionActivityTest {
```
[the remark this answers is not in the document]
Comment (lipefree) [reply app/src/androidTest/java/com/github/studydistractor/sdp/DistractionActivityTest.kt:47]: I can't just delete DistractionActivity or the MapActivity won't work anymore

File: app/src/androidTest/java/com/github/studydistractor/sdp/MainNavigationTest.kt
[the remark this answers is not in the document]
Comment (lipefree) [reply app/src/androidTest/java/com/github/studydistractor/sdp/MainNavigationTest.kt:1]: I don't see the code you are refering to, but my guess is that the correct behavior is `composeRule.onNodeWithTag("distraction-screen__main-container").assertDoesNotExist()` which is in the PR

File: app/src/main/AndroidManifest.xml [hunk 17]
```diff
@@ -31,10 +31,10 @@
             android:exported="false"
             android:label="@string/title_activity_maps" />
         <activity
-            android:name=".procrastinationActivity.DistractionList"
+            android:name=".distraction.DistractionList"
```
[the remark this answers is not in the document]
Comment (lipefree) [reply app/src/main/AndroidManifest.xml:34]: For MapActivity

File: app/src/main/java/com/github/studydistractor/sdp/MainActivity.kt [hunk 18]
```diff
@@ -7,9 +7,19 @@ import androidx.compose.foundation.layout.fillMaxSize
...
+import com.github.studydistractor.sdp.history.HistoryInterface
+import com.github.studydistractor.sdp.distraction.DistractionService
 import com.github.studydistractor.sdp.ui.theme.StudyDistractorTheme
+import dagger.hilt.android.AndroidEntryPoint
+import javax.inject.Inject
 
-class MainActivity : AppCompatActivity() {
+@AndroidEntryPoint(AppCompatActivity::class)
+class MainActivity : Hilt_MainActivity() {
+    @Inject
```
[the remark this answers is not in the document]
Comment (lipefree) [reply app/src/main/java/com/github/studydistractor/sdp/MainActivity.kt:18]: Yes but it should be another task then

File: app/src/main/java/com/github/studydistractor/sdp/distraction/Distraction.kt [hunk 19]
```diff
@@ -38,12 +38,12 @@ data class ProcrastinationActivity(
         return 0
     }
 
-    companion object CREATOR : Parcelable.Creator<ProcrastinationActivity> {
-        override fun createFromParcel(parcel: Parcel): ProcrastinationActivity {
-            return ProcrastinationActivity(parcel)
+    companion object CREATOR : Parcelable.Creator<Distraction> {
```
[the remark this answers is not in the document]
Comment (lipefree) [reply app/src/main/java/com/github/studydistractor/sdp/distraction/Distraction.kt:41]: For MapActivity

File: app/src/main/java/com/github/studydistractor/sdp/distraction/DistractionActivity.kt (hunk 10 above)
Comment (lipefree) [reply app/src/main/java/com/github/studydistractor/sdp/distraction/DistractionActivity.kt:24]: For MapActivity

File: app/src/main/java/com/github/studydistractor/sdp/distraction/DistractionActivity.kt (hunk 11 above)
Comment (lipefree) [reply app/src/main/java/com/github/studydistractor/sdp/distraction/DistractionActivity.kt:30]: MapActivity

File: app/src/main/java/com/github/studydistractor/sdp/distraction/DistractionList.kt (hunk 12 above)
Comment (lipefree) [reply app/src/main/java/com/github/studydistractor/sdp/distraction/DistractionList.kt:26]: MapActivity

File: app/src/main/java/com/github/studydistractor/sdp/distraction/DistractionService.kt (hunk 13 above)
Comment (lipefree) [reply app/src/main/java/com/github/studydistractor/sdp/distraction/DistractionService.kt:9]: Because the signature is not supported (I don't know why) and LazyColumns are used in screens

File: app/src/main/java/com/github/studydistractor/sdp/ui/DistractionListScreen.kt (hunk 15 above)
Comment (lipefree) [reply app/src/main/java/com/github/studydistractor/sdp/ui/DistractionListScreen.kt:43]: It is only used in the list, I don't understand why we should separate them, it is just a helper function to display things

File: app/src/androidTest/java/com/github/studydistractor/sdp/DistractionActivityTest.kt (hunk 16 above)
[the remark this answers is not in the document]
Comment (aizensoosuke) [reply app/src/androidTest/java/com/github/studydistractor/sdp/DistractionActivityTest.kt]: Oh, I get it. Here is what i propose :
1. You do your app as a screen, ignoring the fact that map needs it.
2. You create an activity DistractionActivityWrapper in the package "tempActivityWrappers" that will be temporary and that will only contain that code :
```kotlin
@Inject
    lateinit var distractionService : DistractionService
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContent {
            DistractionScreen
        }
    }
```

And of course it will lack the top and bottom bar, but that will only be temporary. It's fine because we still get all the functionality, with the sacrifice of a screen being less clean (but that workflow needs refactoring anyway), and at the benifit of you having done really clean work :)

File: app/src/main/java/com/github/studydistractor/sdp/maps/MapsActivity.kt (hunk 8 above)
Comment (AthenaPK) [reply app/src/main/java/com/github/studydistractor/sdp/maps/MapsActivity.kt]: Hmm maybe what @aizensoosuke proposed below will work? Because right now, this kind of breaks the functionality of maps, since it takes us to `ProcrastinationActivity`, which at this point is deprecated.

File: app/src/androidTest/java/com/github/studydistractor/sdp/DistractionActivityTest.kt (hunk 16 above)
[the remark this answers is not in the document]
Comment (lipefree) [reply app/src/androidTest/java/com/github/studydistractor/sdp/DistractionActivityTest.kt]: I can't do it like that, to know which distraction to display, I pass around a viewModel. I will do in my way but it will be hacky

**Review by aizensoosuke — APPROVED** [review]
Awesome! Thanks for all the effort, i know i was annoying but it will be worth it imo :)

**Review by bluedrack — APPROVED** [review]
LGTM. Everything discuss as been fixed.
